rawhttp-core/src/main/java/rawhttp/core/body/BodyDecoder.java

```diff
@@ -47,29 +47,27 @@ public BodyDecoder(HttpBodyEncodingRegistry registry, List<String> transferEncod
     }
 
     /**
-     * Create a {@link DecodingOutputStream} that decodes the contents written to it before passing it on
+     * Create a decoder {@link OutputStream} that decodes the contents written to it before passing it on
      * to the given {@link OutputStream}.
      *
      * @param out receiver of decoded contents
      * @return a stream that decodes the bytes written into it, then writes the decoded bytes into another stream
      * @throws IOException if an error occurs while writing to the given stream
      */
-    public DecodingOutputStream decoding(OutputStream out) throws IOException {
+    public OutputStream decoding(OutputStream out) throws IOException {
```

Comment:
Isn't this a pointless API change? Why return a less specific type?

Reply:
the `DecodingOutputStream` is very limiting because it extends `FilterOutputStream` what if i want to add a new XOutputStream that uses `InflaterOutputStream` you wont be able to do that. (unless you add the `InflaterOutputStream` as a variable in the class, but then you lose the point of FilterOutputStream)

plus, as i wrote, the `DecodingOutputStream` is used only for "not closing the clients output stream" (or am i wrong?)
the `finishDecoding()` is only used just the close.
so why not create a OutputStream that just dont closes the clients output stream?  see `UnCloseableOutputStream`
this also fixes a scenario in:
https://github.com/renatoathaydes/rawhttp/blob/65c86749ab18fa0d74f89e8c30be29b083916cfc/rawhttp-core/src/main/java/rawhttp/core/body/encoding/DeflateDecoder.java#L17
where in `InflaterOutputStream` is never closed

Reply:
> the DecodingOutputStream is very limiting because it extends FilterOutputStream what if i want to add a new XOutputStream that uses InflaterOutputStream

You mean this is limiting for the implementator of `HttpMessageDecoder`? I disagree.

Besides making it clear the semantics of the decoder stream being returned, it allows an implementation to explicitly run some code at the end, which is very common for decoders to do as they may need to know when the data is fully consumed.

I wanted to make it very clear that the implementations of `DecodingOutputStream` should not close the underlying stream, while still having a clear hook for implementing cleanup/finalizing the decoding algorithm. This is not a limitation, this is a useful restriction to behaviour implementations are expected to follow.

I see no reason to remove this type and I won't accept any PRs removing it, sorry.

>  where in InflaterOutputStream is never closed

We cannot call `close` on that as it would close the underlying stream. You're probably right that we should be wrapping the outputStream given to it to avoid this problem, please make a separate ticket for that if you think this is a problem (we're flushing the stream, but I'm not sure that's enough).